File: kompartment/src/ode/onestep.js
/**
 * The driver the one-step methods share.
 *
 * An explicit Runge-Kutta pair and a Rosenbrock pair differ in how they take a
 * step and in nothing else that matters to a caller: both produce a new state
 * and an estimate of its error, both carry a dense output over the step, and
 * both want the same things done around that -- a step size chosen from the
 * estimate, an end landed on exactly, the discrete events located, the
 * requested output times read off the interpolant, a state that may not go
 * negative kept that way, and a run that has stopped advancing refused rather
 * than continued. Those are here, once. A method supplies the step.
 *
 * WHAT A METHOD SUPPLIES. `method.setUp(ctx)` is called once and returns an
 * object with:
 *
 *   start(t, y)                          the derivative at the start; returns
 *                                        the evaluations spent
 *   attempt(t, y, h, tnew, ynew, err)    one step of size h into `ynew`, its
 *                                        error vector into `err`; returns
 *                                        {fevals, scale}, with `scale` the
 *                                        factor the weighted error is
 *                                        multiplied by along with |h|
 *   denseAt(tq, out)                     the state inside the step just tried
 *   accept(tnew, ynew, reprojected)      the step was taken; returns fevals
 *   restart(t, y)                        the start state was changed under it
 *                                        (a state put onto its constraint);
 *                                        returns fevals
 *   stats()                              whatever else it counted
 *
 * and the method itself carries `id`, `order` (of the error estimate, which
 * sets the exponent the step is controlled with), `defaultMaxSteps`, and two
 * flags for the constraint: `holdsAtZero`, whether the derivative of a state
 * on its bound is held at zero for it, and `snapsToConstraint`, whether a
 * state within its tolerance of the bound and pushed past it is put onto it.
 *
 * THE CONSTRAINT. A state that may not go negative and would is treated two
 * ways, both here. Its violation joins the error test -- a step that carries
 * it below zero by more than its tolerance is rejected and cut like any other
 * -- and what is left below zero after an accepted step, within tolerance by
 * construction, is projected onto zero and counted. Whether the method also
 * integrates the projected system, holding such a state's derivative at zero,
 * is the method's to say: a Rosenbrock step from the bound does, an explicit
 * pair does not, since its stages would straddle the kink and disagree by the
 * whole jump.
 */

import { locateCrossing } from './events.js';

/** Machine epsilon for a double. */
const EPS = 2 ** -52;

/**
 * The stall guard: how many accepted steps to look back over, and how small a
 * fraction of the run they may have covered before the solver is judged to be
 * pinned rather than merely slow. Measured in elapsed time, not step size: a
 * pinned solver alternates tiny accepted steps with rejected larger ones, so
 * the step size on its own looks healthy.
 */
export const STALL_WINDOW_STEPS = 2000;
export const STALL_SPAN_FRACTION = 1e-9;

/** Steps accepted at the floor in a row before the run is refused. */
const MAX_AT_FLOOR = 20;

/** Safety on the step the error estimate allows; the least and most it may change by. */
const SAFETY = 0.9;
const SHRINK_LEAST = 0.1;
const GROW_MOST = 5;

export class SolverError extends Error {
	constructor(message, t, index) {
		super(message);
		this.name = 'SolverError';
		this.t = t;
		this.index = index;
	}
}

/**
 * The one report for a state or derivative that is not a number, so that the
 * solvers say the same thing about the same fault. `index` is the state that
 * was seen to go, when one is known.
 */
export function nonFiniteError(t, index) {
	const where = index == null ? '' : ` (state ${index})`;
	return new SolverError(
		`The state or its derivative became non-finite at t=${t}${where}. Check for `
		+ `division by zero, a negative base raised to a fractional power, or an `
		+ `initial value that is not a number.`,
		t, index,
	);
}

/** The smallest step that moves the clock at `t`. */
function stepFloor(t) {
	return 16 * EPS * Math.abs(t);
}

/**
 * The derivative of the projected system: where a constrained state is at or
 * below zero and its equation would take it lower, it is held. `<= 0` so a
 * state the projection has just put on zero is held there. The largest
 * derivative held back per state since the last reset is kept for the
 * driver, which judges a step's "held at zero" by it.
 */
function heldDerivative(f, nonNegative) {
	const wrapped = (t, y, out) => {
		f(t, y, out);
		for (let i = 0; i < y.length; i++) {
			if (nonNegative[i] && y[i] <= 0 && out[i] < 0) {
				if (-out[i] > wrapped.push[i]) wrapped.push[i] = -out[i];
				out[i] = 0;
			}
		}
		return out;
	};
	wrapped.push = new Float64Array(nonNegative.length);
	return wrapped;
}

/**
 * @param {object} method  see the file comment
 * @param {(t: number, y: Float64Array, out: Float64Array) => Float64Array} f
 * @param {ArrayLike<number>} tspan  the output times; first and last bound the run
 * @param {Float64Array} y0
 * @param {object} opts  rtol, abstol (number or per state), hmax, hmin, h0,
 *   nonNegative (bool[]), maxSteps, events, jacobian, onAccepted(t, y),
 *   onOutput(t, y) for every requested time as it is passed,
 *   onStep(progress, nsteps, t) -> false to abort
 * @returns {{t: Float64Array, y: Float64Array[], stopped: object|null, stats: object}}
 */
export function integrate(method, f, tspan, y0, opts = {}) {
	const neq = y0.length;
	const npts = tspan.length;
	const t0 = tspan[0];
	const tEnd = tspan[npts - 1];
	const dir = Math.sign(tEnd - t0);
	if (dir === 0) throw new SolverError('Simulation start and end time are equal', t0);
	const span = Math.abs(tEnd - t0);

	const rtol = opts.rtol ?? 1e-3;
	const atol = new Float64Array(neq);
	if (typeof opts.abstol === 'number' || opts.abstol == null) atol.fill(opts.abstol ?? 1e-6);
	else for (let i = 0; i < neq; i++) atol[i] = opts.abstol[i];
	// What the error test divides by, with |y|: the absolute tolerance in
	// units of the relative one.
	const threshold = new Float64Array(neq);
	for (let i = 0; i < neq; i++) threshold[i] = atol[i] / rtol;

	const nonNegative = opts.nonNegative ?? null;
	const maxSteps = opts.maxSteps ?? method.defaultMaxSteps ?? 1e6;
	const maxAtFloor = opts.maxConsecutiveMinStep ?? MAX_AT_FLOOR;
	const hmax = opts.hmax > 0 ? Math.min(opts.hmax, span) : 0.1 * span;
	const hminOpt = opts.hmin > 0 ? opts.hmin : 0;
	const stallWindow = opts.stallWindow ?? STALL_WINDOW_STEPS;
	const stallSpan = STALL_SPAN_FRACTION * span;
	const exponent = 1 / (method.order + 1);

	const rhs = nonNegative && method.holdsAtZero ? heldDerivative(f, nonNegative) : f;
	const stepper = method.setUp({ neq, rhs, f, threshold, atol, rtol, opts });

	// Output: one row per requested time.
	const tout = new Float64Array(npts);
	const yout = new Array(npts);
	let nout = 0;
	const record = (tv, yv) => { tout[nout] = tv; yout[nout] = Float64Array.from(yv); nout++; };

	let t = t0;
	const y = Float64Array.from(y0);
	const ynew = new Float64Array(neq);
	const errVec = new Float64Array(neq);
	const interp = new Float64Array(neq);
	record(t0, y);
	let nextOut = 1;

	let nsteps = 0, nfailed = 0, nfevals = 0, nbelowtol = 0, negative = 0;
	const held = nonNegative && method.holdsAtZero ? new Int32Array(neq) : null;

	// Discrete events: every one is terminal, so the step is cut back to the
	// crossing and the caller restarts from it.
	const events = opts.events ?? null;
	const vL = events ? new Float64Array(events.n) : null;
	const vR = events ? new Float64Array(events.n) : null;
	if (events) events.fun(t0, y, vL);
	let stopped = null;

	nfevals += stepper.start(t, y);

	/** |v| against the larger of the two states and the threshold, at its worst; NaN if any entry is. */
	const weighted = (v, ya, yb) => {
		let worst = 0;
		let at = 0;
		for (let i = 0; i < neq; i++) {
			const e = Math.abs(v[i]) / Math.max(Math.abs(ya[i]), Math.abs(yb[i]), threshold[i]);
			if (!(e >= 0) || !Number.isFinite(yb[i])) return { worst: NaN, at: i };
			if (e > worst) { worst = e; at = i; }
		}
		return { worst, at };
	};

	// --- the first step ----------------------------------------------------------
	// From the size of y' against y, and of y'' out of one explicit Euler
	// trial: a hundredth of the tolerance is asked of the local error a step
	// of the method's order would make.
	let stepSize;
	if (opts.h0 > 0) {
		stepSize = opts.h0;
	} else {
		const f0 = stepper.derivativeAtStart();
		const d0 = weighted(y, y, y).worst / rtol;
		const d1 = weighted(f0, y, y).worst / rtol;
		let guess = d0 < 1e-5 || d1 < 1e-5 ? 1e-6 : 0.01 * (d0 / d1);
		guess = Math.min(guess, hmax);
		const trial = new Float64Array(neq);
		const ftry = new Float64Array(neq);
		for (let i = 0; i < neq; i++) trial[i] = y[i] + dir * guess * f0[i];
		rhs(t0 + dir * guess, trial, ftry);
		nfevals++;
		for (let i = 0; i < neq; i++) ftry[i] -= f0[i];
		let d2 = weighted(ftry, y, y).worst / rtol / guess;
		if (!Number.isFinite(d2)) d2 = 100 * d1;
		const m = Math.max(d1, d2);
		const h1 = m <= 1e-15 ? Math.max(1e-6, guess * 1e-3) : (0.01 / m) ** exponent;
		stepSize = Math.min(100 * guess, h1);
	}
	stepSize = Math.min(hmax, Math.max(stepFloor(t0), hminOpt, stepSize));

	let atFloor = 0;
	let stallStep = 0;
	let stallT = t0;

	for (;;) {
		if (nsteps > maxSteps) {
			throw new SolverError(method.stepBudgetMessage(maxSteps, t), t);
		}

		const hmin = Math.max(stepFloor(t), hminOpt);
		stepSize = Math.min(hmax, Math.max(hmin, stepSize));
		let h = dir * stepSize;
		let last = false;
		if (1.1 * stepSize >= Math.abs(tEnd - t)) {
			h = tEnd - t;
			stepSize = Math.abs(h);
			last = true;
		}

		let err = 0;
		let worstAt = 0;
		let failedOnce = false;
		let restart = false;
		let tnew = t;
		for (;;) {
			tnew = last ? tEnd : t + h;
			h = tnew - t;
			const { fevals, scale } = stepper.attempt(t, y, h, tnew, ynew, errVec);
			nfevals += fevals;
			const w = weighted(errVec, y, ynew);
			err = w.worst * stepSize * scale;
			worstAt = w.at;

			// A stage that is not a number is a failed step, not a fatal one:
			// a stage can overshoot into the square root of a slightly negative
			// inventory, or past the instant a rate stops being defined, and a
			// shorter step lands inside the domain again. When even the
			// smallest step cannot be made to say a number, the run is refused.
			if (!Number.isFinite(err)) {
				nfailed++;
				if (stepSize <= hmin) throw nonFiniteError(t, worstAt);
				failedOnce = true;
				stepSize = Math.max(hmin, 0.1 * stepSize);
				h = dir * stepSize;
				last = false;
				continue;
			}

			// A constrained state below zero by more than its tolerance is an
			// error like any other, and the step that put it there is cut --
			// halved, since the estimate says nothing about where the crossing
			// lies.
			let forConstraint = false;
			if (nonNegative && err <= rtol) {
				let worst = 0;
				for (let i = 0; i < neq; i++) {
					if (nonNegative[i] && ynew[i] < 0) {
						const v = -ynew[i] / threshold[i];
						if (v > worst) { worst = v; worstAt = i; }
					}
				}
				if (worst > rtol) { err = worst; forConstraint = true; }
			}

			if (err <= rtol) { atFloor = 0; break; }

			nfailed++;
			// A state within its tolerance of the bound and pushed past it by
			// every step tried never reaches the bound by halving: each shorter
			// step lands it a little nearer, still positive, and the next tries
			// again. Put onto the bound instead, where the method that holds a
			// state there can hold it, and the step is retried from there. The
			// change is smaller than the error test could tell.
			if (forConstraint && method.snapsToConstraint) {
				let snapped = false;
				for (let i = 0; i < neq; i++) {
					if (nonNegative[i] && ynew[i] < 0 && y[i] > 0 && y[i] <= atol[i]) {
						y[i] = 0;
						snapped = true;
					}
				}
				if (snapped) {
					nfevals += stepper.restart(t, y);
					restart = true;
					break;
				}
			}
			const before = stepSize;
			if (forConstraint || failedOnce) stepSize = Math.max(hmin, 0.5 * stepSize);
			else stepSize = Math.max(hmin, stepSize * Math.max(SHRINK_LEAST, SAFETY * (rtol / err) ** exponent));
			failedOnce = true;
			if (stepSize <= hmin) {
				// Nothing shorter is possible: the step is taken as it is and
				// counted, and a run of them is refused.
				if (++atFloor >= maxAtFloor) {
					throw new SolverError(method.floorMessage(t, hmin, worstAt), t, worstAt);
				}
				stepSize = before;
				nbelowtol++;
				break;
			}
			h = dir * stepSize;
			last = false;
		}
		if (restart) continue;
		nsteps++;

		// What is left below zero is within tolerance -- the error test saw to
		// that -- and is projected onto the bound. Counted, because a
		// projection is a number the method did not compute.
		let reprojected = false;
		if (nonNegative) {
			for (let i = 0; i < neq; i++) {
				if (nonNegative[i] && ynew[i] < 0) { ynew[i] = 0; reprojected = true; negative++; }
				// Held through this step: the derivative held back would have
				// moved the state by more than its tolerance over the step. A
				// projection alone is not a hold.
				if (held && rhs.push[i] * stepSize > atol[i]) held[i]++;
			}
			if (held) rhs.push.fill(0);
		}

		if (nsteps - stallStep >= stallWindow) {
			if (Math.abs(tnew - stallT) < stallSpan) {
				throw new SolverError(method.stallMessage(tnew, stallWindow, stallSpan), tnew);
			}
			stallStep = nsteps;
			stallT = tnew;
		}

		const denseAt = (tq) => {
			stepper.denseAt(tq, interp);
			if (nonNegative) for (let i = 0; i < neq; i++) if (nonNegative[i] && interp[i] < 0) interp[i] = 0;
			return interp;
		};

		if (events) {
			// The event that stopped the previous segment is still sitting on
			// zero at `t0`; the locator is told so and steps past it.
			const hit = locateCrossing(events, t, vL, tnew, ynew, vR, denseAt, t0);
			if (hit) {
				tnew = hit.t;
				ynew.set(denseAt(tnew));
				vR.set(hit.values);
				stopped = { t: tnew, y: Float64Array.from(ynew), which: hit.which };
				last = true;
			}
			vL.set(vR);
		}

		while (nextOut < npts) {
			const tq = tspan[nextOut];
			if (dir * (tnew - tq) < 0) break;
			const at = tq === tnew ? ynew : denseAt(tq);
			record(tq, at);
			if (opts.onOutput) opts.onOutput(tq, at);
			nextOut++;
		}

		if (opts.onAccepted) opts.onAccepted(tnew, ynew);

		if (opts.onStep && (nsteps & 31) === 0) {
			if (opts.onStep(Math.abs(tnew - t0) / span, nsteps, tnew) === false) {
				throw new SolverError('Simulation aborted', tnew);
			}
		}

		if (last) break;

		// The step is taken. Told to the method only now, after the output:
		// its dense output belongs to the step as it was taken, and what the
		// method rotates or re-evaluates for the next step must not reach it.
		nfevals += stepper.accept(tnew, ynew, reprojected);

		// The next step, from this one's error -- only when it went through
		// first time; a step that failed on the way has been cut already.
		if (!failedOnce) {
			const grow = err > 0 ? SAFETY * (rtol / err) ** exponent : GROW_MOST;
			stepSize *= Math.min(GROW_MOST, Math.max(SHRINK_LEAST, grow));
		}

		t = tnew;
		y.set(ynew);
	}

	return {
		t: tout.subarray(0, nout),
		y: yout.slice(0, nout),
		stopped,
		stats: {
			nsteps, nfailed, nfevals, nbelowtol, negative, held,
			solver: method.id,
			...stepper.stats(),
		},
	};
}

File: kompartment/python/kompartment/engine/solvers/onestep.py
"""The driver the one-step methods share (``src/ode/core/onestep.js``).

Step-size control, the non-negativity constraint (held derivatives and the
projection onto zero), the stall guard, events and output are the same for
the Rosenbrock and the Dormand-Prince methods; each method supplies only how
a step is attempted and how the solution is read inside one.
"""

from __future__ import annotations

import math
from typing import Any, Callable, Dict, List, Optional

import numpy as np

from . import SolverError
from . import kernels
from .events import locate_crossing

EPS = 2.0 ** -52
STALL_WINDOW_STEPS = 2000
STALL_SPAN_FRACTION = 1e-9
MAX_AT_FLOOR = 20
SAFETY = 0.9
SHRINK_LEAST = 0.1
GROW_MOST = 5.0


def non_finite_error(t: float, index: Optional[int] = None) -> SolverError:
    where = '' if index is None else f' (state {index})'
    return SolverError('nonfinite', f'The state or its derivative became non-finite at t={t}{where}. Check for '
                                    'division by zero, a negative base raised to a fractional power, or an initial '
                                    'value that is not a number.', t)


def step_floor(t: float) -> float:
    return 16 * EPS * abs(t)


class HeldDerivative:
    """The derivative with a constrained state at zero held there.

    ``f`` hands back a fresh array each call (as ``System.rhs`` does), so it
    is returned as it is, and copied only when something is held.
    """

    def __init__(self, f: Callable[[float, np.ndarray], np.ndarray], non_negative: np.ndarray) -> None:
        self.f = f
        self.mask = non_negative
        self.idx = np.nonzero(non_negative)[0].astype(np.int64)
        self.every = self.idx.size == non_negative.size
        self.push = np.zeros(non_negative.size)
        self._above = kernels.kernel('all_above_zero')

    def __call__(self, t: float, y: np.ndarray) -> np.ndarray:
        out = self.f(t, y)
        idx = self.idx
        # Nothing to hold while every constrained state is above zero, which
        # is nearly always: one pass rather than the whole test.
        if not idx.size:
            return out
        if self._above is not None:
            if self._above(y, idx):
                return out
        elif (y.min() if self.every else y[idx].min()) > 0:
            return out
        hold = self.mask & (y <= 0) & (out < 0)
        if hold.any():
            out = np.array(out, dtype=float)
            np.maximum(self.push, np.where(hold, -out, 0.0), out=self.push)
            out[hold] = 0.0
        return out


def integrate(method: Any, f: Callable[[float, np.ndarray], np.ndarray], tspan: Any, y0: np.ndarray,
              opts: Dict[str, Any]) -> Dict[str, Any]:
    tspan = np.asarray(tspan, dtype=float)
    y0 = np.asarray(y0, dtype=float)
    neq = y0.size
    npts = tspan.size
    t0 = float(tspan[0])
    t_end = float(tspan[-1])
    if t_end == t0:
        raise SolverError('span', 'Simulation start and end time are equal', t0)
    direction = 1.0 if t_end > t0 else -1.0
    span = abs(t_end - t0)
    rtol = float(opts.get('rtol') or 1e-3)
    ab = opts.get('abstol', 1e-6)
    atol = np.full(neq, float(ab)) if np.ndim(ab) == 0 else np.array(ab, dtype=float)
    threshold = atol / rtol
    nn = opts.get('non_negative')
    non_negative = np.array(nn, dtype=bool) if nn is not None and any(nn) else None
    max_steps = opts.get('max_steps') or method.default_max_steps
    hmax = min(opts['hmax'], span) if (opts.get('hmax') or 0) > 0 else 0.1 * span
    hmin_opt = opts['hmin'] if (opts.get('hmin') or 0) > 0 else 0.0
    stall_window = opts.get('stall_window') or STALL_WINDOW_STEPS
    stall_span = STALL_SPAN_FRACTION * span
    exponent = 1 / (method.order + 1)
    rhs = HeldDerivative(f, non_negative) if (non_negative is not None and method.holds_at_zero) else f
    stepper = method.set_up(neq, rhs, f, threshold, atol, rtol, opts)
    tout: List[float] = []
    yout: List[np.ndarray] = []

    def record(tv: float, yv: np.ndarray) -> None:
        tout.append(tv)
        yout.append(np.array(yv, dtype=float))

    t = t0
    y = y0.copy()
    record(t0, y)
    next_out = 1
    nsteps = nfailed = nbelowtol = negative = 0
    nfevals = 0
    held = np.zeros(neq, dtype=np.int64) if isinstance(rhs, HeldDerivative) else None
    events = opts.get('events')
    vl = vr = None
    if events is not None:
        vl = np.array(events.fun(t0, y), dtype=float)
        vr = np.zeros(events.n)
    stopped = None
    nfevals += stepper.start(t, y)

    compiled_norm = kernels.kernel('weighted')
    norm_out = np.zeros(2)
    below = kernels.kernel('any_below_zero')
    nn_idx = np.nonzero(non_negative)[0].astype(np.int64) if non_negative is not None else None

    def any_negative(v: np.ndarray) -> bool:
        """Whether a constrained state is below zero in ``v``."""
        if below is not None:
            return bool(below(v, nn_idx))
        return bool((non_negative & (v < 0)).any())

    def weighted(v: np.ndarray, ya: np.ndarray, yb: np.ndarray) -> tuple:
        if compiled_norm is not None:
            compiled_norm(v, ya, yb, threshold, norm_out)
            return float(norm_out[0]), int(norm_out[1])
        return kernels.weighted_py(v, ya, yb, threshold)

    if (opts.get('h0') or 0) > 0:
        step_size = float(opts['h0'])
    else:
        f0 = stepper.derivative_at_start()
        d0 = weighted(y, y, y)[0] / rtol
        d1 = weighted(f0, y, y)[0] / rtol
        guess = 1e-6 if (d0 < 1e-5 or d1 < 1e-5) else 0.01 * (d0 / d1)
        guess = min(guess, hmax)
        trial = y + direction * guess * f0
        ftry = np.array(rhs(t0 + direction * guess, trial), dtype=float)
        nfevals += 1
        ftry = ftry - f0
        d2 = weighted(ftry, y, y)[0] / rtol / guess
        if not math.isfinite(d2):
            d2 = 100 * d1
        m = max(d1, d2)
        h1 = max(1e-6, guess * 1e-3) if m <= 1e-15 else (0.01 / m) ** exponent
        step_size = min(100 * guess, h1)
    step_size = min(hmax, max(step_floor(t0), hmin_opt, step_size))
    at_floor = 0
    stall_step = 0
    stall_t = t0
    ynew = np.zeros(neq)
    err_vec = np.zeros(neq)
    while True:
        if nsteps > max_steps:
            raise SolverError('steps', method.step_budget_message(max_steps, t), t)
        hmin = max(step_floor(t), hmin_opt)
        step_size = min(hmax, max(hmin, step_size))
        h = direction * step_size
        last = False
        if 1.1 * step_size >= abs(t_end - t):
            h = t_end - t
            step_size = abs(h)
            last = True
        err = 0.0
        worst_at = 0
        failed_once = False
        restart = False
        tnew = t
        while True:
            tnew = t_end if last else t + h
            h = tnew - t
            fevals, scale = stepper.attempt(t, y, h, tnew, ynew, err_vec)
            nfevals += fevals
            worst, worst_at = weighted(err_vec, y, ynew)
            err = worst * step_size * scale
            if not math.isfinite(err):
                nfailed += 1
                if step_size <= hmin:
                    raise non_finite_error(t, worst_at)
                failed_once = True
                step_size = max(hmin, 0.1 * step_size)
                h = direction * step_size
                last = False
                continue
            for_constraint = False
            if non_negative is not None and err <= rtol and any_negative(ynew):
                v = np.where(non_negative & (ynew < 0), -ynew / threshold, 0.0)
                at = int(np.argmax(v))
                if v[at] > 0 and v[at] > rtol:
                    err = float(v[at])
                    worst_at = at
                    for_constraint = True
            if err <= rtol:
                at_floor = 0
                break
            nfailed += 1
            if for_constraint and method.snaps_to_constraint:
                snap = non_negative & (ynew < 0) & (y > 0) & (y <= atol)
                if snap.any():
                    y[snap] = 0.0
                    nfevals += stepper.restart(t, y)
                    restart = True
                    break
            before = step_size
            if for_constraint or failed_once:
                step_size = max(hmin, 0.5 * step_size)
            else:
                step_size = max(hmin, step_size * max(SHRINK_LEAST, SAFETY * (rtol / err) ** exponent))
            failed_once = True
            if step_size <= hmin:
                at_floor += 1
                if at_floor >= MAX_AT_FLOOR:
                    raise SolverError('tolerance', method.floor_message(t, hmin, worst_at), t)
                step_size = before
                nbelowtol += 1
                break
            h = direction * step_size
            last = False
        if restart:
            continue
        nsteps += 1
        reprojected = False
        if non_negative is not None:
            if any_negative(ynew):
                neg = non_negative & (ynew < 0)
                ynew[neg] = 0.0
                reprojected = True
                negative += int(neg.sum())
            if held is not None:
                held += (rhs.push * step_size > atol).astype(np.int64)  # type: ignore[union-attr]
                rhs.push[:] = 0.0  # type: ignore[union-attr]
        if nsteps - stall_step >= stall_window:
            if abs(tnew - stall_t) < stall_span:
                raise SolverError('stalled', method.stall_message(tnew, stall_window, stall_span), tnew)
            stall_step = nsteps
            stall_t = tnew

        def dense_at(tq: float) -> np.ndarray:
            out = stepper.dense_at(tq)
            if non_negative is not None:
                out = np.where(non_negative & (out < 0), 0.0, out)
            return out

        if events is not None:
            hit = locate_crossing(events, t, vl, tnew, ynew, vr, dense_at, t0)
            if hit is not None:
                tnew = hit['t']
                ynew[:] = dense_at(tnew)
                vr[:] = hit['values']
                stopped = {'t': tnew, 'y': ynew.copy(), 'which': hit['which']}
                last = True
            vl[:] = vr
        while next_out < npts:
            tq = float(tspan[next_out])
            if direction * (tnew - tq) < 0:
                break
            at_v = ynew if tq == tnew else dense_at(tq)
            record(tq, at_v)
            if opts.get('on_output') is not None:
                opts['on_output'](tq, at_v)
            next_out += 1
        if opts.get('on_accepted') is not None:
            opts['on_accepted'](tnew, ynew)
        if opts.get('on_step') is not None and (nsteps & 31) == 0:
            if opts['on_step'](abs(tnew - t0) / span, nsteps, tnew) is False:
                raise SolverError('aborted', 'Simulation aborted', tnew)
        if last:
            break
        nfevals += stepper.accept(tnew, ynew, reprojected)
        if not failed_once:
            grow = SAFETY * (rtol / err) ** exponent if err > 0 else GROW_MOST
            step_size *= min(GROW_MOST, max(SHRINK_LEAST, grow))
        t = tnew
        y = ynew.copy()
    stats = {'nsteps': nsteps, 'nfailed': nfailed, 'nfevals': nfevals, 'nbelowtol': nbelowtol, 'negative': negative,
             'held': held, 'solver': method.id, 'points': len(tout)}
    stats.update(stepper.stats())
    return {'t': np.array(tout), 'y': yout, 'stopped': stopped, 'stats': stats}
